scramjetorg/transform-hub
# KM3 persistent sequence

This PR is used for testing and future fixing overflowing of runner socket.

Expected behaviour:
After starting instance without end consumer, writing to instance more data than inner buffers allow should cause pause event on flowing chunks.

Actual behaviour:
Writing more data than inner buffer should allow, are buffered in event queue by socket. Causing drain event right after pause, and never passing pause to generator. Additionaly creating more events in queue leads to:
`MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [Socket]. Use emitter.setMaxListeners() to increase limit`

Expected fix:
connection between runner socket and host socket server should behave like regular stream, causing pause/resume events on inner buffer overflow.

How to test on regular instance:
1. run sth
2. deploy sequence
3. start sequence
4. Write some big file to created instance- This should be paused if no consumer is attached, and not accepted by socket

How to test on persistent sequence:
1.run sth:  `SCRAMJET_DEVELOPMENT=1 yarn start:dev --runtime-adapter=process`
2.  In `cd src/transform-hub/bdd/data/sequences/persistentSeq` run: 
`yarn predeploy`
`yarn deploy`
3. `curl -X POST -H "content-type: application/json" -d '{"id":"testTopic", "content-type":"text/plain", "persistentSequence": "persistentSeq"}' 127.0.0.1:8000/api/v1/topics`
4. `si topic send testTopic /transform-hub/bdd/data/sequences/persistentSeq.tar.gz`

Places to look:
STH socket server creating connection to runner: packages/host/src/lib/socket-server.ts
Piping connection from socket server to topics etc: packages/host/src/lib/csi-controller.ts:455
Creating socket for runner to connect with STH: packages/runner/src/host-client.ts
 packages/runner/src/runner.ts:580

Steps to take:
simplify runner to check default behafiuor of socket:
begining of simplified implementation is here: bdd/data/sequences/persistentSeq/example.ts
